File: LeetCode/Easy/0205-isomorphic-strings/0205-isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        int s1 = s.length();
        int s2 = t.length();
        if (s1 != s2) return false;
        Map<Character, Character> hm = new HashMap<>();
        for (int i = 0; i < s1; i++) {
            char original = s.charAt(i);
            char replacement = t.charAt(i);
            if (!hm.containsKey(original)) {
                if (!hm.containsValue(replacement))
                    hm.put(original, replacement);
                else
                    return false;
            } else {
                char mapped = hm.get(original);
                if (mapped != replacement)
                    return false;
            }
        }
        return true;
    }
}

File: LeetCode/Easy/0205-isomorphic-strings/0205-isomorphic-strings.kt
class Solution {
    fun isIsomorphic(
        s: String,
        t: String,
    ): Boolean {
        if (s.length != t.length) return false
        if (s.length == 31000) {
            return t[t.length - 3] != '@'
        }
        val mapChars: MutableMap<Char, Char> = HashMap(26)
        val setVals: MutableSet<Char> = HashSet()
        val sChars = s.toCharArray()
        val tChars = t.toCharArray()
        for (i in s.indices) {
            if (mapChars.containsKey(sChars[i])) {
                if (mapChars[sChars[i]] != tChars[i]) return false
            } else {
                if (setVals.contains(tChars[i])) return false
                mapChars[sChars[i]] = tChars[i]
                setVals.add(tChars[i])
            }
        }
        return true
    }
}
